[feature] 현재 위치 저장 및 공유 위한 데이터 레이어 구현

miller198 opened the pull request:
## #️⃣연관된 이슈

> closes #108

## 📝작업 내용 및 코드

1. LocalDataSource
```kotlin
class LocalDataSourceImpl : LocalDataSource {
    
    private var _userId: String? = null
    private var _currentLocation: MutableStateFlow<Location?> = MutableStateFlow(null)

    override val userId: String?
        get() = _userId

    override val currentLocation: StateFlow<Location?>
        get() = _currentLocation

    override fun saveUserId(userId: String) {
        _userId = userId
    }

    override suspend fun saveCurrentLocation(location: Location) {
        _currentLocation.emit(location)
    }
}
```
2. LocalRepository
```kotlin
class LocalRepositoryImpl @Inject constructor(
    private val localDataSource: LocalDataSource
) : LocalRepository {

    override val userId get() = localDataSource.userId
    override val currentLocation get() = localDataSource.currentLocation

    override suspend fun saveUserId(userId: String) {
        localDataSource.saveUserId(userId)
    }

    override suspend fun saveCurrentLocation(location: Location) {
        localDataSource.saveCurrentLocation(location)
    }
}
```
3. MapViewModel에서 5m 이상 이동해야 LocalDataSource의 현재 위치 값 변경하고 주변 픽 fetch 해오도록 

### 이하 11.19

### MusicRoadMarker

- 네이버지도 SDK에서 기본 제공하는 클래스인 Marker()와 Pick 타입 데이터를 함께 캡슐화한 Wrapper 클래스
- 우리 앱에서 지도의 마커는 모두 Pick 데이터를 하나씩 담고 있으므로 하나의 클래스로 묶어서 사용하는게 편리하다고 생각함
- Marker 클래스를 상속받아서 마커 자체를 새로 만드려고 했으나 final이라 상속이 안됨
- Overlay는 상속받을수있긴한데 너무 어려워 보여서 포기

```kotlin
class MusicRoadMarker(
    val userId: String = "",
    val pick: Pick,
) {
    val position get() = marker.position
    val map get() = marker.map
    private val marker = Marker()

    operator fun invoke() = marker

    fun clearMap() {
        marker.map = null
    }

    fun loadMarkerImage(context: Context, map: NaverMap, onMarkerClick: () -> Unit) {
        val markerIconView = MarkerIconView(context).apply {
            setPaintColor(Blue.toArgb())
        }

        markerIconView.loadImage(pick.song.getImageUrlWithSize(REQUEST_IMAGE_SIZE)) {
            marker.position = LatLng(pick.location.latitude, pick.location.longitude)
            marker.icon = OverlayImage.fromView(markerIconView)
            marker.setOnClickListener {
                onMarkerClick()
                true
            }
            marker.map = map
        }
    }

    fun toggleSizeByClick(context: Context, isClicked: Boolean) {
        val defaultIconWidth = marker.icon.getIntrinsicWidth(context)
        val defaultIconHeight = marker.icon.getIntrinsicHeight(context)

        marker.width =
            if (isClicked) (defaultIconWidth * MARKER_SCALE).toInt() else defaultIconWidth
        marker.height =
            if (isClicked) (defaultIconHeight * MARKER_SCALE).toInt() else defaultIconHeight
    }

    companion object {
        val REQUEST_IMAGE_SIZE = Size(300, 300)
        const val MARKER_SCALE = 1.5
    }
}
```

- 생성자에서 Pick 데이터를 받아서 만들어짐
- userid는 추후 사용할것같아서 만들어두고 현재 쓰이지 않음 - 이 마커를 만든 사람 혹은 Pick 만든 유저를 의미
- 기존 NaverMap.kt의 createMarker() 내에서 하던 동작들을 옮겨와서 메소드로 만듦
- 또한 toggleSizeByClick 같은 marker 데이터에 대한 조작도 함께 캡슐화

### 위치 포함 픽 정보 등록

현재 위치 포함해서 Firestore에 픽을 등록하고 PickDetail 화면으로 전환

여기서 뒤로가기를 누르면 메인지도로 넘어가고 위치에 픽이 등록된것을 확인가능

[참고] (https://[file.notion.so/f/f/74aecf81-0d2a-4c29-9052-a7585287d611/a3eafb36-3e7f-4442-8b67-2b1333d09b48/untitled.gif?table=block&id=143ea5e5-fc17-8002-89d0-c8194264ab50&spaceId=74aecf81-0d2a-4c29-9052-a7585287d611&expirationTimestamp=1732111200000&signature=0JJ5Gmd9HWIgQiuBftR_ZIME1c1pGHKoiPui7ZawYls&downloadName=untitled.gif](https://file.notion.so/f/f/74aecf81-0d2a-4c29-9052-a7585287d611/a3eafb36-3e7f-4442-8b67-2b1333d09b48/untitled.gif?table=block&id=143ea5e5-fc17-8002-89d0-c8194264ab50&spaceId=74aecf81-0d2a-4c29-9052-a7585287d611&expirationTimestamp=1732111200000&signature=0JJ5Gmd9HWIgQiuBftR_ZIME1c1pGHKoiPui7ZawYls&downloadName=untitled.gif))


```kotlin
// NavGraph
composable(CreatePickDestinations.CREATE_PICK_ROUTE) {
    val parentEntry = remember(it) {
        navController.getBackStackEntry(CreatePickDestinations.SEARCH_ROUTE)
    }
    CreatePickScreen(
        createPickViewModel = hiltViewModel<CreatePickViewModel>(parentEntry),
        onBackClick = {
            navController.popBackStack()
        },
        onCreateClick = { pickId ->
            navigationActions.navigateToPickDetail(pickId)
        }
    )
}
```

- CreatePickScreen 에 onCreateClick 함수로 navigationActions.navigateToPickDetail(pickId) 전달

```kotlin
// CreatePickScreen
CreatePickDisplay(
    song = song,
    comment = comment.value,
    dynamicBackgroundColor = dynamicBackgroundColor,
    dynamicOnBackgroundColor = dynamicOnBackgroundColor,
    onBackClick = {
        createPickViewModel.resetComment()
        onBackClick()
    },
    onCreateClick = {
        createPickViewModel.createPick(onCreateClick)
    },
    onCommentChange = createPickViewModel::onCommentChange
)
```

- CreatePickScreen 에서 이를 createPickViewModel.createPick() 호출에 사용함

```kotlin
// CreatePickViewModel
fun createPick(
    onSuccess: (String) -> Unit
) {
    _selectedSong?.let { song ->
        viewModelScope.launch {
            val musicVideoUrl = searchMusicVideoById(song.id)

            if (curLocation == DEFAULT_LOCATION) {
                /* TODO: DEFAULT 인 경우 -> LocalDataSource 위치 데이터 못 불러옴 */
            }

            /* 등록 결과 - pick ID 담긴 Result */
            /* FIXME : createdBy 임시값 */
            val createResult = createPickUseCase(
                Pick(
                    id = "",
                    song = song,
                    comment = _comment.value,
                    createdAt = "",
                    createdBy = "",
                    location = LocationPoint(curLocation.latitude, curLocation.longitude),
                    musicVideoUrl = musicVideoUrl
                )
            )

            createResult.onSuccess { pickId ->
                onSuccess(pickId)
            }.onFailure {
                /* TODO: 실패처리 */
                Log.d("CreatePickViewModel", createResult.exceptionOrNull()?.message.toString())
            }
        }
    }
}
```

- CreatePickViewModel 에선 onSuccess파라미터로 이를 받아서 Firestore에 등록 동작이 성공했을때 호출
- 결과적으로 등록 성공하면 Detail 스크린으로 넘어가는 것임

### 거리계산

InfoWindow 거리계산

![image](https://github.com/user-attachments/assets/a954cbff-e566-464b-ad6d-ad060ff00b15)

![image](https://github.com/user-attachments/assets/7ddaa935-b122-4374-b47b-7fede8107843)


```kotlin

fun InfoWindow(
    pick: Pick,
    navigateToPick: (String) -> Unit,
    calculateDistance: (Double, Double) -> String
) 

		...
		
		Text(
		    text = "${calculateDistance(pick.location.latitude, pick.location.longitude)}m", // TODO pick.distance 값
		    style = MaterialTheme.typography.bodyMedium.copy(Gray)
		)
```

- InfoWindow의 거리 텍스트 다음과 같이 수정

```kotlin
 fun calculateDistance(
    from: Location = curLocation.value,
    lat: Double,
    lng: Double,
): Double {
    val location = Location("").apply {
        latitude = lat
        longitude = lng
    }
    return from.distanceTo(location).toDouble()
}
```

- MapViewModel에 거리계산 함수 추가
- InfoWindow의 거리 텍스트에서 이 함수를 호출하여 사용하는 중

## 💬리뷰 요구사항(선택)

miller198 committed:
[refactor] PickLocation -> LocationPoint

miller198 committed:
[feature] LocalDataSource, Repository 추가

uid, pickid, 현재 좌표 저장 위한 Local 데이터 관련 클래스

miller198 committed:
Merge branch 'develop' into feature/local-repository

miller198 committed:
[refactor] LocalDataSource 삭제

착각해서 만든것 삭제

miller198 committed:
[refactor] LocalDataSource 다시 추가

LocalRepository의 책임을 더 명확하기 만들기 위해

miller198 committed:
[refactor] selectedPick 삭제

selectedPick은 로컬DataSource에서 저장할 필요가 없다

miller198 committed:
[feature] 지도에서 픽 5미터 간격 불러오기

LocalRepository와 DataSource, flow 이용해 5미터 이상 이동했을때만 주변 픽 불러오는 기능

miller198 committed:
[refactor] MapViewModel 인스턴스 생성 위치 변경

액티비티에서 가지고 있을 필요 없으므로 MainNavGraph로 선언 및 생성위치 변경

miller198 committed:
[feature] SearchMusicScreen curLocation collect 추가

데이터레이어에 저장된 위치 UI까지 불러오는것 확인

miller198 committed:
[refactor] CreatePickViewModel curLocation stateIn 삭제

검색화면 전환시 DEFAULT_LOCATION 까지 두번 받아오는 문제 수정

miller198 committed:
[refactor] MapScreen 주변 픽 불러오는 조건 수정

MapScreen 에서 DEFAULT 위치 받아올때는 주변픽정보 불러오지 않도록 수정

meanjoo commented on app/src/main/java/com/squirtles/musicroad/create/SearchMusicScreen.kt:
```diff
@@ -75,6 +77,11 @@ fun SearchMusicScreen(
     val searchText by createPickViewModel.searchText.collectAsStateWithLifecycle()
     val searchResult by createPickViewModel.searchResult.collectAsStateWithLifecycle()
     val isSearching by createPickViewModel.isSearching.collectAsStateWithLifecycle()
+    val curLocation by createPickViewModel.curLocation.collectAsStateWithLifecycle()
+
+    LaunchedEffect(curLocation) {
+        Log.d("SearchMusicScreen", curLocation.toString())
+    }
```
검색 화면에서 현재 위치 값을 collect하는 이유가 있나요??

단지 값을 확인해보기 위해서 작성한 거라면 없어도 되지 않을까요? 현재 위치 값을 collect 하여 수행하는 작업이 있다면 어떤 것인지 궁금합니다.

miller198 replied:
collect 해서 UI업데이트같은 동작은 하지 않습니다.
생각해보니 UI에서 이걸 알고 있을 필요도 없군요. 단순히 CreatePickViewModel에서 가지고 있기만 하면 되네요.
CreatePickViewModel에서 저장하고 있던 curLocation은 이후 CreatePickScreen에서 등록버튼을 눌러서 실행되는 
createPick()에서만 사용하면 될 것 같습니다.
CreatePickViewModel에서 curLocation을 데이터 레이어에서 가져올때 일반 Location 타입으로 저장만 하도록 수정하겠습니다

meanjoo commented on app/src/main/java/com/squirtles/musicroad/map/MapViewModel.kt:
```diff
@@ -2,46 +2,74 @@ package com.squirtles.musicroad.map
+        if (curLocationPoint == DEFAULT_LOCATION) {
+            saveCurLocation(location)
+        } else {
+            if (calculateDistance(location, curLocationPoint) > 5.0) {
+                saveCurLocation(location)
```
위치가 5m 변할 때마다 현재 위치를 저장하고 현재 위치가 변할 때 주변 픽을 불러오는 것은 좋은 것 같습니다!

등록할 때도 현재 위치 값을 쓰게 될 텐데 하나 궁금증이 있습니다.
현재 위치 값을 5m가 변할 때마다 저장한다면 그림처럼 초록색 사람이 왼쪽으로 이동하고 있을 때, 빨간색 점에서 위치 업데이트가 일어났으면 초록색 점에서 위치 업데이트가 다시 일어날 것입니다. 픽을 등록할 때 현재 위치 값을 사용할 텐데 그럼 사용자가 주황색 점에서 음악 등록 버튼 클릭을 누르면 빨간색 점에 마커가 생기게 되나요?
![image](https://github.com/user-attachments/assets/09fa873b-db53-4988-ac30-a7ec3a9939cd)

yuni-ju commented on app/src/main/java/com/squirtles/musicroad/create/CreatePickViewModel.kt:
```diff
@@ -1,22 +1,26 @@
 
 @HiltViewModel
 class CreatePickViewModel @Inject constructor(
+    private val fetchLocationUseCase: FetchLocationUseCase,
```
단순히 내부에서 값을 그대로 사용하려면 private val을 제거하고 fetchLocationUseCase: FetchLocationUseCase로 써도 될 것 같습니다

miller198 replied:
맞습니다 반영하겠습니다

miller198 committed:
[refactor] 리뷰 반영

검색화면에서 필요 없는 curLocation 삭제, CreatePickViewModel에서 flow가 아닌 일반 Location 타입으로 curLocation 저장

Co-Authored-By: minju <88606886+meanjoo@users.noreply.github.com>

yuni-ju commented on app/src/main/java/com/squirtles/musicroad/pick/DetailPickScreen.kt:
```diff
@@ -59,18 +58,18 @@ import com.squirtles.musicroad.ui.theme.White
 
 @Composable
 fun DetailPickScreen(
-    pickId: String,
```
제 개인적 의견으로는 DetailPick화면은 여러 경로로 편리하게 접근하고 재사용이 용이하도록 pickId를 파라미터로 전달 받아서 쓰는 것은 어떤가요? viewModel 내부에서 savedStateHandle로 전달받아 사용하는 것은 확인했으나 detail/pickId의 경로로 id값만 받아오는 것이 더 직관적일 것 같아보입니다! 

savedStateHandle로 전달 받도록 하신 이유가 궁금합니다.

miller198 replied:
saveStateHandle로 전해진 데이터를 받아올 수 있는지 확인하다 남겨진 코드입니다 
수정해서 반영하겠습니다!

miller198 committed:
Merge branch 'develop' into feature/#108-local-repository

miller198 committed:
[refactor] 리뷰 반영

saveStateHandle사용X, 지도에서 등록 버튼 누르면 LocalDataSource의 위치 데이터 변경하도록 수정

miller198 committed:
[refactor] 디버그 Log 제거

miller198 committed:
[feature] MusicRoadMarker 추가

marker와 pick을 함께 묶은 클래스

miller198 committed:
[feature] 위치 정보 포함 픽 등록 및 화면 전환

Firestore에 현재위치를 담은 픽을 등록하고 등록 이후 DetailPick화면으로 전환

miller198 committed:
[feature] InfoWindow 거리계산

yuni-ju commented on app/src/main/java/com/squirtles/musicroad/map/MapScreen.kt:
```diff
@@ -70,17 +73,26 @@ fun MapScreen(
                     modifier = Modifier.padding(bottom = 16.dp),
                     onFavoriteClick = onFavoriteClick,
-                    onCenterClick = onCenterClick,
+                    onCenterClick = {
+                        onCenterClick()
+                        mapViewModel.onCenterButtonClick()
+                    },
```
지도에서 위치를 불러오기 전 버튼을 클릭하면 현재 위치가 null일 것입니다.
mapViewModel.onCenterButtonClick() 내부에서 null인 경우 값을 업데이트하지 않고 return을 하지만,
onCenterClick()은 동작해서 등록 화면으로 넘어가는 상태인 것 같습니다! 

이때, 픽을 등록하면 default 위치로 등록이 되긴하지만 결국 사용자가 확인할 수 없는 위치라 의미가 없을 것 같다는 생각이 듭니다. 

혹시 현재 위치가 null이 아닐 때만 onCenterClick 기능이 동작하도록 하는 것은 어떨까요?

miller198 replied:
이러면 CreatePickViewModel 에서 LocalDataSource의 위치데이터 불러왔을때 null 일 경우가 없겠네요 좋습니다

yuni-ju reviewed (commented):
PickState에서 previous, current가 가리키는 픽이 없는 상태나 Location이 null일 때는 이를 명확하게 파악할 수 있도록 비어있는 문자열이나 0.0 값보다 nullable하게 쓰는 것은 어떨까요? 

DEFAULT_LOCATION을 썼을 때 그냥 null을 썼을 때보다 해당 값이 의미있게 쓰이거나 좀 더 장점이 있는지 궁금합니다!

meanjoo commented on app/src/main/java/com/squirtles/musicroad/map/MapViewModel.kt:
```diff
@@ -1,95 +1,133 @@
+        lat: Double,
+        lng: Double,
+    ): Double {
+        val location = Location("").apply {
```
큰 의미가 있는 건 아니지만 Location의 provider로 `Location("pickLocation")`과 같이 의미있는 이름을 붙여주는 건 어떨까요?

![image](https://github.com/user-attachments/assets/d921b9d0-3339-43e6-89a4-27e1249da8a4)

생성자에서 이 위치와 연관된 provider 이름을 인자로 받는다고 하여 생각해본 것이었습니다!

miller198 replied:
넵 좋습니다

meanjoo commented on app/src/main/java/com/squirtles/musicroad/map/MapViewModel.kt:
```diff
@@ -1,95 +1,133 @@
 
-            Log.d("MapViewModel", pick.toString())
+    /* 미터 단위 거리 계산 */
+    fun calculateDistance(from: Location = curLocation.value, to: Location): Double =
+        from.distanceTo(to).toDouble()
```
이 함수는 뷰모델 안의 updateCurLocation 안에서만 사용되는 것 같아서 private으로 변경해도 될 것 같습니다.

개인적으로는 함수 안에서 어떤 함수 하나만을 호출하고 있는데 굳이 이걸 함수로 분리해야했을까? 하는 생각이 듭니다. 함수 분리를 하지 않은 코드도 충분히 어떤 동작을 하는지 명확하다고 생각하는데 함수로 따로 분리하신 이유가 궁금합니다.
```kotlin
// in updateCurLocation()

// 현재 코드
if (calculateDistance(to = location) > 5.0) {
    saveCurLocation(location)
}

// 함수 분리를 하지 않았을 때
if (curLocation.value.distanceTo(location) > 5.0) {
    saveCurLocation(location)
}
```

miller198 replied:
처음에 이 함수를 MapScreen의 BottomNavigation에서 인포윈도우에서 거리계산할때도 쓰지않을까 해서 public상태인 분리된 함수로 만들었습니다

인포윈도우 거리 계산도 구현한 지금 상황에서 거리계산 함수가 인포윈도우에서 사용하는 것과 updateCurLocation() 안에서 사용하는것 두개가 있는데 이 두개를 하나로 통합하는게 낫겠네요

meanjoo commented on app/src/main/java/com/squirtles/musicroad/map/MapViewModel.kt:
```diff
@@ -1,95 +1,133 @@
+            latitude = lat
+            longitude = lng
         }
+        return from.distanceTo(location).toDouble()
```
현재 이 함수를 호출하는 곳은 한 군데 밖에 없고 아래와 같이 호출하고 있습니다.
`mapViewModel.calculateDistance(lat = lat, lng = lng).toInt().toString()`

from을 안 넘기고 있어서 기본 값인 curLocation.value가 사용되고 있습니다. 거리 계산 함수는 현재 위치를 기준으로만 사용될 것 같은데 from을 없애고 `return curLocation.value.distanceTo(location).toDouble()`로 변경해도 되지 않을까요? from을 받으면서 거리 계산 함수를 사용하는 곳이 있을지 궁금합니다.

+) distanceTo()는 반환형이 Float인데 왜 Double형으로 반환하는지도 궁금합니다. 사용하는 곳에서는 결국 Int로 바꿔서 사용하고 있는데 그냥 Float으로 반환해도 상관없지 않나 하는 생각이 들어서 여쭤봅니다.

miller198 replied:
코틀린 실수는 기본이 Double형이라 제가 그걸 선호합니다 어차피 Int로 바꿀거면 그게 나을 수 있는데 파라미터로 받는 lat, lng는 Double 타입이라 리턴 타입도 맞추는게 좋겠네요
from 파라미터는 이걸 사용하는 곳이 없어도 함수 자체가 더 유연하게 사용되도록 파라미터는 남겨두고 기본값을 사용하는게 좋다고 생각했습니다. 일반적으로 거리계산에는 두개의 좌표가 필요하니까요

meanjoo replied:
아하 그냥 왜 Double로 반환하나 궁금한 거였습니다!

함수 자체를 유연하게 만들기 위해서 from 파라미터를 남겨두고 기본값을 설정한 후 이를 사용하는 것은 이해했습니다.
함수를 더 유연하게 만드는 것도 좋지만 만약 프로젝트 내에서 전혀 from을 사용할 일이 없고 항상 from이 현재 위치로 고정된다면 calculateDistanceFromCurLocation()와 같이 아예 함수명 자체를 변경해도 나쁘진 않을 것 같네욥!

miller198 commented:
## 리뷰 반영사항

**1. curLocation -> lastLocation 변경**
LocalDataSource에 저장되는 위치 변수는
5미터 이상 유저 위치가 변할때마다 저장되는 변수이고, 
이는 실시간 현재 위치가 아니므로 의미를 명확히 하기위해 변경하였습니다

대신 BottomNavigtaion에서 등록버튼 눌렀을때 MapViewModel의 saveCurLocationForced (이전 이름 onCenterButtonClick) 을 실행하면 현재 유저의 실시간 위치를 저장하는데, 저장되는 변수 명을 realtimeLocation -> curLocation으로 변경하였습니다

**2. 바텀 내비게이션 파일 분리**
컴포저블 함수가 길어져서 파일을 분리하였습니다

**3. 바텀 내비게이션 중앙버튼 클릭 시 마지막 위치값이 없으면 화면 전환 되지 않도록 수정**
```kotlin
@Composable
fun MapScreen(
    mapViewModel: MapViewModel,
    onFavoriteClick: () -> Unit,
    onCenterClick: () -> Unit,
    onSettingClick: () -> Unit,
    onInfoWindowClick: (String) -> Unit
) {
    val pickCount by mapViewModel.pickCount.collectAsStateWithLifecycle()
    val pickMarkers by mapViewModel.pickMarkers.collectAsStateWithLifecycle()
    val selectedPickState by mapViewModel.selectedPickState.collectAsStateWithLifecycle()
    val lastLocation by mapViewModel.lastLocation.collectAsStateWithLifecycle()
```
MapScreen에서 mapViewModel 의 lastLocation 관찰

```kotlin
BottomNavigation(
    modifier = Modifier.padding(bottom = 16.dp),
    onFavoriteClick = onFavoriteClick,
    onCenterClick = {
        // 마지막으로 저장된 위치값이 없으면 화면전환 X
        if (lastLocation != null) {
            onCenterClick()
            mapViewModel.saveCurLocationForced()
        }
    },
    onSettingClick = onSettingClick
)
```
BottomNavigation의 중앙 버튼에 대해 
lastLocation 이 null 이면 ( = 아직 유저 위치 불러와 저장하지 못한 상태) 동작되지 않도록 변경

```kotlin
fun createPick(
        onSuccess: (String) -> Unit
    ) {
        _selectedSong?.let { song ->
            viewModelScope.launch {
                val musicVideoUrl = searchMusicVideoById(song.id)

                if (lastLocation == null) {
                    /* TODO: DEFAULT 인 경우 -> LocalDataSource 위치 데이터 못 불러옴 */
                    return@launch
                }
```
이렇게 한다면 위의 CreatePickViewModel::createPick() 에서 lastLocation null체크를 안해도 될 것 같은데 어떻게 생각하세요?

CreatePickViewModel은 바텀내비게이션 등록버튼을 눌렀을때 SearchMusicScreen으로 전환되면서 그 인스턴스가 생성되는데,
SearchMusicScreen으로의 전환은 LocalDataSource의 lastLocation != null 일 경우에만 전환되기때문에 CreatePickViewModel에서 lastLocation == null 일 경우가 없을 것 같습니다


**4. MapViewModel calucate함수 하나로 통합 및 km 단위로 수정**
```kotlin
fun calculateDistance(
    lat: Double,
    lng: Double,
    from: Location? = lastLocation.value,
): Double {
    return from?.let {
        val location = Location("pickLocation").apply {
            latitude = lat
            longitude = lng
        }

        val distanceInMeters = from.distanceTo(location).toDouble()
        val distanceInKm = distanceInMeters / 1000.0

        String.format(Locale.getDefault(), "%.1f", distanceInKm).toDouble()
    } ?: -1.0
}
```
만약 지도를 축소해서 100km 바깥의 픽을 보면 100000m 이런식으로 나오기 때문에
km 단위로 소수점 첫째자리까지만 반올림되어서 나오게 수정

그리고 -1.0이 리턴되는 경우는 lastLocation.value 가 null 일때 이므로,
이때는 인포윈도우에 거리 표현을 할수 없기 때문에 인포윈도우에선 빈 문자열이 나오도록 수정 하였습니다

```kotlin
calculateDistance = { lat, lng ->
    val distance = mapViewModel.calculateDistance(lat = lat, lng = lng)
    if (distance == -1.0) "" else "${distance}km"
}
```


**5. PickState의 프로퍼티 및 lastLocation Nullable하게 변경**
지적해주신대로 의미 상 해당 데이터가 없으면 null 인게 맞는듯하여 변경하였습니다.

meanjoo commented:
3에 대한 제 생각입니다. lastLocation이 null이 아닐 때만 화면이 전환되면 createPick에서 널 체크를 하지 않아도 될 것 같긴 합니다.
혹시 이런 것도 가능할까요? lastLocation이 null일 때는 바텀 내비게이션 바의 중앙 버튼의 클릭을 막고 색상을 회색 같이 아직 상호작용 할 수 없을 것 같은 색으로 변경하는 것도 가능한가요?

miller198 committed:
[refactor] 리뷰반영

바텀내비게이션파일분리, curLocation -> LastLocation이름 변경, 등록버튼 동작 조건 추가, MapViewModel의 거리계산 함수 통합 및 km 단위 변경, https://github.com/boostcampwm-2024/and06-musicroad/pull/109#issuecomment-2487447799

miller198 commented:
비상 비상
![image](https://github.com/user-attachments/assets/47a1b2a0-cf5e-4a43-9f37-f880406a0525)

~~바텀 버튼 위치가 이상해졌는데 도와주세요~~

해결...ㅎㅎㅎ

miller198 committed:
[refactor] 등록 버튼 비활성화 시 색상 변경

바텀내비게이션 패딩 수정, lastLocation == null 일때 등록버튼 색상을 회색으로 변경

miller198 commented:
@meanjoo @yuni-ju 

[untitled.webm](https://github.com/user-attachments/assets/3012c516-be20-42de-92ec-eb0582dd1277)


이렇게 하면 될까요?

miller198 committed:
[fix] resetSelectedPickState 오류 수정

resetSelectedPickState 함수에서 Pickstate의 current를 "" 이 아닌 null 로 저장하도록 수정

meanjoo commented:
@miller198 오!👍👍👍 저는 좋은데 윤겸님 의견도 들어봅시당

yuni-ju commented:
저도 좋은 것 같습니다! 👍

meanjoo commented on app/src/main/java/com/squirtles/musicroad/map/MapViewModel.kt:
```diff
@@ -1,95 +1,122 @@
 
-            Log.d("MapViewModel", pick.toString())
-        }
+            val distanceInMeters = from.distanceTo(location).toDouble()
+            val distanceInKm = distanceInMeters / 1000.0
+
+            String.format(Locale.getDefault(), "%.1f", distanceInKm).toDouble()
+        } ?: -1.0
```
이렇게 하면 인포윈도우에서 거리가 표시될 때 무조건 km 단위로 표시되더라구요.
저는 1km 이상일 때는 km 단위, 그 미만일 때는 m 단위로 나왔으면 좋겠는데 어떻게 생각하시나요??

만약 저렇게 하려면 calculateDistance()에서는 미터 단위로 반환하고 값을 사용하는 인포윈도우 쪽에서 체크하고 변환하는 건 어떤가요? 이렇게 한다면 코드는 이런 식으로 바꿀 수 있을 것 같아요.
```kotlin
// in MapViewModel.kt
fun calculateDistance(
    lat: Double,
    lng: Double,
    from: Location? = lastLocation.value,
): Double {
    return from?.let {
        val location = Location("pickLocation").apply {
            latitude = lat
            longitude = lng
        }

        from.distanceTo(location).toDouble()
    } ?: -1.0
}

// in MapScreen.kt
InfoWindow(
    pick,
    navigateToPick = { pickId ->
        onInfoWindowClick(pickId)
    },
    calculateDistance = { lat, lng ->
        val distance = mapViewModel.calculateDistance(lat = lat, lng = lng)
        when {
            distance >= 1000.0 -> String.format(Locale.getDefault(), "%.1fkm", distance / 1000.0)
            distance >= 0 -> String.format(Locale.getDefault(), "%.0fm", distance)
            else -> ""
        }
    }
)
```

miller198 replied:
네 이 코드 사용해서 수정해볼게요

meanjoo commented on app/src/main/java/com/squirtles/musicroad/map/MapViewModel.kt:
```diff
@@ -1,95 +1,122 @@
-            _curLocation.value = location
+        _currentLocation = location
+
+        if (lastLocation.value == null
+            || calculateDistance(location.latitude, location.longitude) > 5.0
```
calculateDistance()에 from을 안 넘겨서 기본 값인 lastLocation.value와 거리를 계산하고 있습니다. calculateDistance()에서 lastLocation.value가 null이면 -1을 반환해서 lastLocation.value가 널인 거를 체크할 필요가 없지 않을까요?
반환값이 > 5.0인지만 체크해도 될 것 같습니다.

miller198 replied:
~~그렇네요 반영하겠습니다!~~

lastLocation.value == null 일 경우는 lastLocation에 아무 값도 저장되지 않은 초기 상태라는 말이라서 이건 체크해야합니다 
안하니까 등록버튼이 계속 회색이네요

meanjoo replied:
아하 이해완료했습니다

miller198 committed:
[refactor] 인포윈도우 거리 표시 수정

1km 이하 시 m 단위, 넘으면 km 단위

Co-Authored-By: minju <88606886+meanjoo@users.noreply.github.com>

miller198 committed:
[refactor] saveCurLocationForced() 에 viewModelScope 제거

Co-Authored-By: minju <88606886+meanjoo@users.noreply.github.com>
